presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDirectSelectiveStreamReader.java

```diff
@@ -0,0 +1,525 @@
+            }
+            streamPosition++;
+        }
+        outputPositionCount = positionCount;
+        outputPositions = positions;
+        return streamPosition;
+    }
+
+    private int readWithFilter(int[] positions, int positionCount)
+            throws IOException
+    {
+        int streamPosition = 0;
+        allNulls = false;
+        offsetVector = new int[positionCount + 1];
+
+        int offsetVectorIndex = 0;
+        offsetVector[offsetVectorIndex] = 0;
+        boolean isCharType = streamDescriptor.getOrcType().getOrcTypeKind() == OrcType.OrcTypeKind.CHAR;
+        int maxCodePointCount = getMaxCodePointCount(streamDescriptor.getOrcType());
+        int failedLengthCheckSize = 0;
+
+        for (int i = 0; i < positionCount; i++) {
+            int position = positions[i];
+            if (position > streamPosition) {
+                skipData(streamPosition, position);
+                streamPosition = position;
+            }
+
+            if (isNullVector != null && isNullVector[position]) {
+                if ((nonDeterministicFilter && filter.testNull()) || nullsAllowed) {
+                    int prev = offsetVector[offsetVectorIndex];
+                    offsetVector[++offsetVectorIndex] = prev;
+                    if (outputRequired) {
+                        nulls[outputPositionCount] = true;
+                    }
+                    outputPositions[outputPositionCount] = position;
+                    outputPositionCount++;
+                }
+            }
+            else {
+                int off = offsetVector[offsetVectorIndex];
+                int currentLength = lengthVector[lengthIndex];
+                //the dataStream can be null here and lengthVector is not null
+                if (dataStream != null) {
+                    if (filter.testLength(currentLength)) {
+                        //skip failed lengthCheckSize as dataStream has to be advanced for those
+                        dataStream.skip(failedLengthCheckSize);
+                        failedLengthCheckSize = 0;
+                        dataStream.next(data, off, off + currentLength);
+                        if (filter.testBytes(data, off, currentLength)) {
+                            if (outputRequired) {
+                                int truncatedLength = computeTruncatedLength(slice, off, currentLength, maxCodePointCount, isCharType);
+                                Verify.verify(truncatedLength >= 0);
+                                offsetVector[++offsetVectorIndex] = off + truncatedLength;
+                            }
+                            outputPositions[outputPositionCount] = position;
+                            outputPositionCount++;
+                        }
+                    }
+                    else {
+                        failedLengthCheckSize += currentLength;
+                    }
+                }
+                lengthIndex++;
+            }
+
+            streamPosition++;
```

Comment:
@bhhari I think it is necessary to check if `failedLengthCheckSize > 0` and if so call `dataStream.skip(failedLengthCheckSize);`. Otherwise, if positions at the end of this batch fail length check, the data stream won't advance properly and the next `read` will break.